### PR TITLE
Issue #97: Argo's use of command and args fields do not follow Kubern…

### PR DESCRIPTION
**Pull request by sytianhe**
…etes

### CODE REVIEW

File: platform/source/tools/container_outer_executor.py [hunk 1]
```diff
@@ -521,25 +513,25 @@ def _pre_run(self):
...
-                    token = match_obj.group(1)
-                    if token and token not in self._encrypted_strings:
-                        try:
-                            self._encrypted_strings[token] = axops_client.decrypt(token=token, repo_name=self._repo)
-                        except Exception as e:
-                            logger.exception("Failed to decrypt %s", token)
-                            self._gen_user_command_with_failed_decryption(token, error_messages=str(e))
-                            self._write_executor_sh()
-                            return False
+        # logger.info("Start decrypting.")
```
Comment (abhinavdas) [inline platform/source/tools/container_outer_executor.py]: why comment this out? Remove the code if not needed...

File: platform/source/tools/container_outer_executor.py [hunk 2]
```diff
@@ -735,69 +727,78 @@ def _split_cmd(cmdString):
...
+
+        # According to https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/
+        # 1. If you do not supply command or args for a Container, the defaults defined in the Docker image are used.
+        # 2. If you supply a command but no args for a Container, only the supplied command is used. The default EntryPoint and the default Cmd defined in the Docker image are ignored.
+        # 3. If you supply only args for a Container, the default Entrypoint defined in the Docker image is run with the args that you supplied.
+        # 4. If you supply a command and args, the default Entrypoint and the default Cmd defined in the Docker image are ignored. Your command is run with your args.
+
+        # Condition #1
+        if command is None and args is None:
+            # process entrypoint_in_image
```
Comment (abhinavdas) [inline platform/source/tools/container_outer_executor.py]: I think we should move this code out to some code in docker module

File: platform/source/tools/container_outer_executor.py [hunk 3]
```diff
@@ -735,69 +727,78 @@ def _split_cmd(cmdString):
...
+            if entrypoint_in_image and isinstance(entrypoint_in_image, string_types):
+                entrypoint_in_image = ["/bin/sh", "-c", entrypoint_in_image]
+            if entrypoint_in_image and not isinstance(entrypoint_in_image, list):
```
Comment (abhinavdas) [inline platform/source/tools/container_outer_executor.py]: Am I not seeing the else case here where entrypoint_in_image is a list....

File: platform/source/tools/container_outer_executor.py [hunk 4]
```diff
@@ -735,69 +727,78 @@ def _split_cmd(cmdString):
...
+                logger.error("<env>: bad entrypoint_in_image %s", entrypoint_in_image)
+                entrypoint_in_image = None
+            # process command_in_image
+            if command_in_image and isinstance(command_in_image, string_types):
+                command_in_image = ["/bin/sh", "-c", command_in_image]
+            if command_in_image and not isinstance(command_in_image, list):
```
Comment (abhinavdas) [inline platform/source/tools/container_outer_executor.py]: `else isintance(command_in_image, list)`

File: platform/source/tools/container_outer_executor.py [hunk 5]
```diff
@@ -735,69 +727,78 @@ def _split_cmd(cmdString):
...
-            elif command_in_image:
-                return command_in_image
-            else:
+                logger.error("<env>: bad command %s, expecting a list", command)
                 return None
 
+        # Condition #3
+        if command is None and args is not None:
+            # process entrypoint_in_image
+            if entrypoint_in_image and isinstance(entrypoint_in_image, string_types):
```
Comment (abhinavdas) [inline platform/source/tools/container_outer_executor.py]: This is common code with condition 1. Should be shared...

File: platform/source/tools/container_outer_executor.py [hunk 6]
```diff
@@ -735,69 +727,78 @@ def _split_cmd(cmdString):
...
+                entrypoint_in_image = ["/bin/sh", "-c", entrypoint_in_image]
+            if entrypoint_in_image and not isinstance(entrypoint_in_image, list):
+                logger.error("<env>: bad entrypoint_in_image %s", entrypoint_in_image)
+                entrypoint_in_image = None
+
+            if isinstance(args, list):
```
Comment (abhinavdas) [inline platform/source/tools/container_outer_executor.py]: args will always be a list or None...

File: platform/source/tools/container_outer_executor.py [hunk 7]
```diff
@@ -735,69 +727,78 @@ def _split_cmd(cmdString):
...
+                if entrypoint_in_image is None:
+                    return args
+                else:
+                    return entrypoint_in_image + args
+            else:
```
Comment (abhinavdas) [inline platform/source/tools/container_outer_executor.py]: Not needed

File: platform/source/tools/container_outer_executor.py (hunk 1 above)
Comment (sytianhe) [reply platform/source/tools/container_outer_executor.py]: @pratulw and @edlee2121 would like to keep the secret management v1. Will make another change regarding to this.

**Comment by jessesuen** [comment]
In general, I am okay with the changes since I saw the end-to-end test passing for my specific issue. It would be preferable to have a unit test testing all the possible combinations...

**Pull request merged**